extensions/common/vault/vault-hashicorp/src/main/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultTokenRenewTask.java

```diff
@@ -0,0 +1,131 @@
+
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.system.ExecutorInstrumentation;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This task implements the Hashicorp Vault token renewal mechanism.
+ * To ensure that this task is really cancelled, call the stop method before program shut down.
+ */
+public class HashicorpVaultTokenRenewTask {
+
+    private static final String INITIAL_TOKEN_LOOK_UP_ERR_MSG_FORMAT = "Initial token look up failed with reason: %s";
+    private static final String INITIAL_TOKEN_RENEW_ERR_MSG_FORMAT = "Initial token renewal failed with reason: %s";
+    private static final String SCHEDULED_TOKEN_RENEWAL_ERR_MSG_FORMAT = "Scheduled token renewal failed: %s";
+
+    @NotNull
+    private final ExecutorInstrumentation executorInstrumentation;
+    @NotNull
+    private final HashicorpVaultClient client;
+    @NotNull
+    private final Monitor monitor;
+    private final long renewBuffer;
+    private ScheduledExecutorService scheduledExecutorService;
+    private Future<?> tokenRenewTask;
+    private final AtomicBoolean isRunning = new AtomicBoolean(false);
+
+    public HashicorpVaultTokenRenewTask(@NotNull ExecutorInstrumentation executorInstrumentation,
+                                        @NotNull HashicorpVaultClient client,
+                                        long renewBuffer,
+                                        @NotNull Monitor monitor) {
+        this.executorInstrumentation = executorInstrumentation;
+        this.client = client;
+        this.renewBuffer = renewBuffer;
+        this.monitor = monitor;
+    }
+
+    /**
+     * Starts the scheduled token renewal.
+     * Runs asynchronously.
+     */
+    public void start() {
+        if (!isRunning()) {
+            scheduledExecutorService = executorInstrumentation.instrument(Executors.newSingleThreadScheduledExecutor(), HashicorpVaultExtension.NAME);
+            scheduledExecutorService.execute(this::initialize);
+            isRunning.set(true);
+        }
+    }
+
+    /**
+     * Stops the scheduled token renewal. Running tasks will be interrupted.
+     */
+    public void stop() {
+        if (isRunning()) {
+            if (tokenRenewTask != null) {
+                tokenRenewTask.cancel(true);
+            }
+            scheduledExecutorService.shutdownNow();
+            isRunning.set(false);
+        }
+    }
+
+    public boolean isRunning() {
+        return isRunning.get();
+    }
+
+    private void initialize() {
+        var tokenLookUpResult = client.lookUpToken();
+
+        if (tokenLookUpResult.succeeded()) {
+            var isRenewable = tokenLookUpResult.getContent();
+
+            if (isRenewable) {
+                renewToken(INITIAL_TOKEN_RENEW_ERR_MSG_FORMAT);
```

Comment:
shouldn't this schedule the renewal instead of calling it immediately?

Reply:
I believe the EDC should instantly look up and renew the token

1) you don't know the actual ttl of the token (the token could be really old already) -> I wouldn't want to schedule a renewal with a fictional ttl which might not fit the actual ttl. You could go around this by scheduling with 0s but then the log message will spew out that something went wrong with the scheduling instead of the "initialization" logic.
2) The token might be invalid from the start and you'd want to know that instantly
3) It was also requested this way & I think it makes sense to do so (or at the very least it doesn't hurt)

Reply:
sounds ok to me but that's not what's stated in the PR description:
>  If the token is renewable a scheduled one-shot task is submitted. The task executes after a delay defined as TTL - VAULT_TOKEN_RENEW_BUFFER

Reply:
updated the text